README.md
# Better API Fetch


```
#login with your personal username/password
npn login 
# set the organization locally
npm config set scope forgaia
## publish the packages
npm publish --access public
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.1.3](https://github.com/forgaia/fetch/compare/v1.1.2...v1.1.3) (2020-03-17)

### [1.1.2](https://github.com/forgaia/fetch/compare/v1.1.1...v1.1.2) (2020-03-12)

### [1.1.1](https://github.com/forgaia/fetch/compare/v1.1.0...v1.1.1) (2020-03-12)

## [1.1.0](https://github.com/forgaia/fetch/compare/v1.0.148...v1.1.0) (2020-03-12)


### Features

* **ci:** ci ([b67867f](https://github.com/forgaia/fetch/commit/b67867f1a7ade9e2762391169ab4396edc956fb3))

### [1.0.150](https://github.com/forgaia/fetch/compare/v1.0.148...v1.0.150) (2020-03-12)

### [1.0.149](https://github.com/forgaia/fetch/compare/v1.0.148...v1.0.149) (2020-03-12)

### [1.0.148](https://github.com/forgaia/fetch/compare/v1.0.147...v1.0.148) (2020-03-12)

### [1.0.147](https://github.com/forgaia/fetch/compare/v1.0.146...v1.0.147) (2020-03-12)

### [1.0.146](https://github.com/forgaia/fetch/compare/v1.0.145...v1.0.146) (2020-03-12)

### [1.0.145](https://github.com/forgaia/fetch/compare/v1.0.144...v1.0.145) (2020-03-12)

### [1.0.144](https://github.com/forgaia/fetch/compare/v1.0.143...v1.0.144) (2020-03-12)

### [1.0.143](https://github.com/forgaia/fetch/compare/v1.0.142...v1.0.143) (2020-03-12)

### [1.0.142](https://github.com/forgaia/fetch/compare/v1.0.141...v1.0.142) (2020-03-12)

### [1.0.141](https://github.com/forgaia/fetch/compare/v1.0.140...v1.0.141) (2020-03-12)

### [1.0.140](https://github.com/forgaia/fetch/compare/v1.0.139...v1.0.140) (2020-03-12)

### [1.0.139](https://github.com/forgaia/fetch/compare/v1.0.138...v1.0.139) (2020-03-12)

### [1.0.138](https://github.com/forgaia/fetch/compare/v1.0.137...v1.0.138) (2020-03-12)

### [1.0.137](https://github.com/forgaia/fetch/compare/v1.0.136...v1.0.137) (2020-03-12)

### [1.0.136](https://github.com/forgaia/fetch/compare/v1.0.135...v1.0.136) (2020-03-12)

### [1.0.135](https://github.com/forgaia/fetch/compare/v1.0.134...v1.0.135) (2020-03-12)

### [1.0.134](https://github.com/forgaia/fetch/compare/v1.0.133...v1.0.134) (2020-03-12)

### [1.0.133](https://github.com/forgaia/fetch/compare/v1.0.132...v1.0.133) (2020-03-12)

### [1.0.132](https://github.com/forgaia/fetch/compare/v1.0.131...v1.0.132) (2020-03-12)

### [1.0.131](https://github.com/forgaia/fetch/compare/v1.0.130...v1.0.131) (2020-03-12)

### [1.0.130](https://github.com/forgaia/fetch/compare/v1.0.129...v1.0.130) (2020-03-12)

### [1.0.129](https://github.com/forgaia/fetch/compare/v1.0.128...v1.0.129) (2020-03-12)

### [1.0.128](https://github.com/forgaia/fetch/compare/v1.0.127...v1.0.128) (2020-03-12)

### [1.0.127](https://github.com/forgaia/fetch/compare/v1.0.126...v1.0.127) (2020-03-12)

### [1.0.126](https://github.com/forgaia/fetch/compare/v1.0.125...v1.0.126) (2020-03-12)

### [1.0.125](https://github.com/forgaia/fetch/compare/v1.0.124...v1.0.125) (2020-03-12)

### [1.0.124](https://github.com/forgaia/fetch/compare/v1.0.123...v1.0.124) (2020-03-12)

### [1.0.123](https://github.com/forgaia/fetch/compare/v1.0.122...v1.0.123) (2020-03-12)

### [1.0.122](https://github.com/forgaia/fetch/compare/v1.0.121...v1.0.122) (2020-03-12)

### [1.0.121](https://github.com/forgaia/fetch/compare/v1.0.120...v1.0.121) (2020-03-12)

### [1.0.120](https://github.com/forgaia/fetch/compare/v1.0.119...v1.0.120) (2020-03-12)

### [1.0.119](https://github.com/forgaia/fetch/compare/v1.0.118...v1.0.119) (2020-03-12)

### [1.0.118](https://github.com/forgaia/fetch/compare/v1.0.117...v1.0.118) (2020-03-12)

### [1.0.117](https://github.com/forgaia/fetch/compare/v1.0.116...v1.0.117) (2020-03-12)

### [1.0.116](https://github.com/forgaia/fetch/compare/v1.0.115...v1.0.116) (2020-03-12)

### [1.0.115](https://github.com/forgaia/fetch/compare/v1.0.114...v1.0.115) (2020-03-12)

### [1.0.114](https://github.com/forgaia/fetch/compare/v1.0.113...v1.0.114) (2020-03-12)

### [1.0.113](https://github.com/forgaia/fetch/compare/v1.0.112...v1.0.113) (2020-03-12)

### [1.0.112](https://github.com/forgaia/fetch/compare/v1.0.111...v1.0.112) (2020-03-12)

### [1.0.111](https://github.com/forgaia/fetch/compare/v1.0.110...v1.0.111) (2020-03-12)

### [1.0.110](https://github.com/forgaia/fetch/compare/v1.0.109...v1.0.110) (2020-03-12)

### [1.0.109](https://github.com/forgaia/fetch/compare/v1.0.108...v1.0.109) (2020-03-12)

### [1.0.108](https://github.com/forgaia/fetch/compare/v1.0.107...v1.0.108) (2020-03-12)

### [1.0.107](https://github.com/forgaia/fetch/compare/v1.0.106...v1.0.107) (2020-03-12)

### [1.0.106](https://github.com/forgaia/fetch/compare/v1.0.105...v1.0.106) (2020-03-12)

### [1.0.105](https://github.com/forgaia/fetch/compare/v1.0.104...v1.0.105) (2020-03-12)

### [1.0.104](https://github.com/forgaia/fetch/compare/v1.0.103...v1.0.104) (2020-03-12)

### [1.0.103](https://github.com/forgaia/fetch/compare/v1.0.102...v1.0.103) (2020-03-12)

### [1.0.102](https://github.com/forgaia/fetch/compare/v1.0.101...v1.0.102) (2020-03-12)

### [1.0.101](https://github.com/forgaia/fetch/compare/v1.0.100...v1.0.101) (2020-03-12)

### [1.0.100](https://github.com/forgaia/fetch/compare/v1.0.99...v1.0.100) (2020-03-12)

### [1.0.99](https://github.com/forgaia/fetch/compare/v1.0.98...v1.0.99) (2020-03-12)

### [1.0.98](https://github.com/forgaia/fetch/compare/v1.0.96...v1.0.98) (2020-03-12)

### [1.0.97](https://github.com/forgaia/fetch/compare/v1.0.96...v1.0.97) (2020-03-12)

### [1.0.96](https://github.com/forgaia/fetch/compare/v1.0.95...v1.0.96) (2020-03-12)

### [1.0.95](https://github.com/forgaia/fetch/compare/v1.0.94...v1.0.95) (2020-03-12)

### [1.0.94](https://github.com/forgaia/fetch/compare/v1.0.93...v1.0.94) (2020-03-12)

### [1.0.93](https://github.com/forgaia/fetch/compare/v1.0.92...v1.0.93) (2020-03-12)

### [1.0.92](https://github.com/forgaia/fetch/compare/v1.0.91...v1.0.92) (2020-03-12)


test/HttpError.test.ts
import HttpError, { errorMessages } from '../src/HttpError';

describe('HttpError', () => {
  const error = new HttpError(500, '/url', 'error', 'Something went wrong.');
  it('inherit Error class', () => {
    expect(error).toBeInstanceOf(Error);
  });

  it('code is set', () => {
    expect(error.code).toEqual(500);
  });

  it('name is set', () => {
    expect(error.name).toEqual('HTTP Error');
  });
  it('Override correct friendlyMessage for known status', () => {
    expect(error.friendlyMessage).toEqual('Something went wrong.');
  });
  it('has fallback friendlyMessage for unknown status', () => {
    const error2 = new HttpError(999, '/url', 'error');
    expect(error2.friendlyMessage).toEqual('Something went wrong.');
  });

  it('has correct friendlyMessage for known status 500', () => {
    const error2 = new HttpError(500, '/url', 'error');
    expect(error2.friendlyMessage).toEqual(errorMessages[500]);
  });
  it('has correct friendlyMessage for known status 401', () => {
    const error2 = new HttpError(401, '/url', 'error');
    expect(error2.friendlyMessage).toEqual(errorMessages[401]);
  });
  it('has correct friendlyMessage for known status 404', () => {
    const error2 = new HttpError(404, '/url', 'error');
    expect(error2.friendlyMessage).toEqual(errorMessages[404]);
  });
  it('has correct friendlyMessage for known status 444', () => {
    const error2 = new HttpError(444, '/url', 'error');
    expect(error2.friendlyMessage).toEqual(errorMessages[444]);
  });
});


src/FetchAPI.js
import HttpError from './HttpError';
export default class FetchAPI {
    static setShouldMockAllRequests(shouldMockAllRequests) {
        this.shouldMockAllRequests = shouldMockAllRequests;
    }
    static setBaseUrl(baseUrl) {
        this.baseUrl = baseUrl;
    }
    static setDefaultHeaders(headers) {
        Object.keys(headers).forEach((key) => {
            FetchAPI.defaultHeaders[key] = headers[key];
        });
    }
    static setAuthorizationToken(token) {
        FetchAPI.setDefaultHeaders({ Authorization: `Bearer ${token}` });
    }
    static set setDefaultRequestInfo(reqInfo) {
        Object.keys(reqInfo).forEach((key) => {
            // @ts-ignore
            FetchAPI.defaultRequestInfo[key] = reqInfo[key];
        });
    }
    static serializeQueryString(params) {
        const urlSearchParams = Object.keys(params).map(key => {
            // Encode objects and array to json. (used for cellIds and drillParams.)
            const value = typeof params[key] === 'object' ? JSON.stringify(params[key]) : params[key];
            return [encodeURIComponent(key), encodeURIComponent(value)].join('=');
        });
        return urlSearchParams.join('&');
    }
    static getUrl(req) {
        if (req.params) {
            return `${this.baseUrl}${req.url}?${this.serializeQueryString(req.params)}`;
        }
        return `${this.baseUrl}${req.url}`;
    }
    static getHeaders(req) {
        const headers = new Headers(this.defaultHeaders);
        if (this.shouldMockAllRequests || req.shouldMockRequest) {
            headers.set('x-mock-response', 'yes');
        }
        return headers;
    }
    static fetch(req) {
        const url = this.getUrl(req);
        return fetch(url, {
            ...this.defaultRequestInfo,
            ...req,
            headers: this.getHeaders(req)
        })
            .then(response => {
            if (response.ok) {
                return response.json();
            }
            return response.json().then(json => {
                if (json.status_code) {
                    // The API provides custom response on errors. As shows in the example below. This code handles this situation by creating custom HttpError Class which allows sentry.io and AppDynamics to capture the error information.
                    // json example: {
                    // status_code: 404;
                    // title: "Resource Not Found"
                    // message: "Requested resource is unavailable, it's either been moved or deleted."
                    // }
                    return Promise.reject(new HttpError(json.status_code, response.url, json.message, json.title));
                }
                return Promise.reject(new HttpError(response.status, response.url, response.statusText));
            });
        })
            .catch(error => {
            throw new HttpError(error.code || 500, url, error.message || '', error.friendlyMessage);
        });
    }
}
FetchAPI.shouldMockAllRequests = false;
FetchAPI.defaultHeaders = {
    'Access-Control-Allow-Origin': '*',
    'Content-Type': 'application/json'
};
FetchAPI.defaultRequestInfo = {
    method: 'GET'
};


src/FetchAPI.ts
import HttpError from './HttpError';

export interface RequestConfig {
  /**
   * Returns the URL of request as a string.
   */
  readonly url: string;
  /**
   * Returns request's HTTP method, which is "GET" by default.
   */
  readonly method?: string;
  /**
   * List if allowed query string params for this endpoint.
   * All allowed keys must exists inside the the provided params array.
   */
  readonly allowedParams: string[];
  /**
   * Whether to add the mock HTTP header to the Request or not.
   */
  readonly shouldMockRequest?: boolean;
}

interface RequestParams extends Partial<Omit<Request, 'body' | 'url' | 'method'>>, RequestConfig {
  readonly params?: Record<string, any>;
  readonly body?: string;
}

export default class FetchAPI {
  private static shouldMockAllRequests: boolean = false;

  private static baseUrl: string;

  /**
   * Treat some http code as valid response
   * For Example: in PM Project we treat 422 (Validation error) as valid response)
   */
  private static whitelistedHTTPCodes: number[] = [];

  private static defaultHeaders: Record<string, string> = {
    'Access-Control-Allow-Origin': '*',
    'Content-Type': 'application/json'
  };

  private static defaultRequestInfo: Partial<RequestParams> = {
    method: 'GET'
  };

  public static setWhitelistedHTTPCodes(errorCodes: number[]) {
    this.whitelistedHTTPCodes = errorCodes;
  }

  public static setShouldMockAllRequests(shouldMockAllRequests: boolean) {
    this.shouldMockAllRequests = shouldMockAllRequests;
  }

  public static setBaseUrl(baseUrl: string) {
    this.baseUrl = baseUrl;
  }

  public static setDefaultHeaders(headers: Record<string, string>) {
    Object.keys(headers).forEach((key: string) => {
      FetchAPI.defaultHeaders[key] = headers[key];
    });
  }

  public static setAuthorizationToken(token: string) {
    FetchAPI.setDefaultHeaders({ Authorization: `Bearer ${token}` });
  }

  public static setDefaultRequestInfo(reqInfo: Partial<Request>) {
    Object.keys(reqInfo).forEach((key: string) => {
      // @ts-ignore
      FetchAPI.defaultRequestInfo[key] = reqInfo[key];
    });
  }

  /**
   * @param params
   * @param allowedParams
   */
  private static serializeQueryString(params: any, allowedParams: string[]): string {
    return allowedParams
      .map(key => {
        // Encode objects and array to json. (used for cellIds and drillParams.)
        const value = typeof params[key] === 'object' ? JSON.stringify(params[key]) : params[key];
        return [encodeURIComponent(key), encodeURIComponent(value)].join('=');
      })
      .join('&');
  }

  private static getUrl(req: RequestParams): string {
    if (req.params) {
      const qs = this.serializeQueryString(req.params, req.allowedParams);
      return `${this.baseUrl}${req.url}${qs ? `?${qs}` : ''}`;
    }
    return `${this.baseUrl}${req.url}`;
  }

  private static getHeaders(req: RequestParams): Headers {
    const headers = new Headers(this.defaultHeaders);

    if (this.shouldMockAllRequests || req.shouldMockRequest) {
      headers.set('x-mock-response', 'yes');
    }

    return headers;
  }

  public static fetch<T = any>(req: RequestParams): Promise<T> {
    const url = this.getUrl(req);

    return fetch(url, {
      ...this.defaultRequestInfo,
      ...req,
      headers: this.getHeaders(req)
    })
      .then(response => {
        if (response.ok || this.whitelistedHTTPCodes.includes(response.status)) {
          return response.json();
        }

        return response.json().then(json => {
          if (json.status_code) {
            // The API provides custom response on errors. As shows in the example below. This code handles this situation by creating custom HttpError Class which allows sentry.io and AppDynamics to capture the error information.
            // json example: {
            // status_code: 404;
            // title: "Resource Not Found"
            // message: "Requested resource is unavailable, it's either been moved or deleted."
            // }
            return Promise.reject(new HttpError(json.status_code, response.url, json.message, json.title));
          }
          return Promise.reject(new HttpError(response.status, response.url, response.statusText));
        });
      })
      .catch(error => {
        throw new HttpError(error.code || 500, url, error.message || '', error.friendlyMessage);
      });
  }
}


src/HttpError.js
export const errorMessages = {
    401: 'Unauthorized, Your token has expired and needs to be renews.',
    404: 'Not Found',
    444: 'Resource Not Found',
    500: 'Sorry, something went wrong.'
};
export default class HttpError extends Error {
    /**
     * @param {number} httpCode
     * @param {string} url
     * @param {string} statusText
     * @param {string} friendlyMessage
     */
    constructor(httpCode, url, statusText, friendlyMessage) {
        super();
        this.url = undefined;
        this.code = undefined;
        this.friendlyMessage = undefined;
        this.name = 'HTTP Error';
        this.code = httpCode;
        this.url = url;
        this.message = statusText || '';
        this.friendlyMessage = friendlyMessage || errorMessages[this.code] || 'Something went wrong.';
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, HttpError);
        }
    }
}


src/FetchAPI.d.ts
interface RequestParams extends Partial<Omit<Request, 'body'>> {
    readonly url: string;
    readonly params?: Record<string, any>;
    readonly shouldMockRequest?: boolean;
    readonly body?: string;
}
export default class FetchAPI {
    private static shouldMockAllRequests;
    private static baseUrl;
    private static defaultHeaders;
    private static defaultRequestInfo;
    static setShouldMockAllRequests(shouldMockAllRequests: boolean): void;
    static setBaseUrl(baseUrl: string): void;
    static setDefaultHeaders(headers: Record<string, string>): void;
    static setAuthorizationToken(token: string): void;
    static set setDefaultRequestInfo(reqInfo: Partial<Request>);
    private static serializeQueryString;
    private static getUrl;
    private static getHeaders;
    static fetch<T = any>(req: RequestParams): Promise<T>;
}
export {};


src/HttpError.ts
export const errorMessages: Record<number, string> = {
  401: 'Unauthorized, Your token has expired and needs to be renews.',
  404: 'Not Found',
  444: 'Resource Not Found',
  500: 'Sorry, something went wrong.'
};

export default class HttpError extends Error {
  public url?: string = undefined;

  public code?: number = undefined;

  public friendlyMessage?: string = undefined;

  /**
   * @param {number} httpCode
   * @param {string} url
   * @param {string} statusText
   * @param {string} friendlyMessage
   */
  constructor(httpCode: number, url: string, statusText: string, friendlyMessage?: string) {
    super();
    this.name = 'HTTP Error';
    this.code = httpCode;
    this.url = url;
    this.message = statusText || '';
    this.friendlyMessage = friendlyMessage || errorMessages[this.code] || 'Something went wrong.';
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, HttpError);
    }
  }
}
